services-core/src/gas_price/linear_interpolation.rs

```diff
@@ -0,0 +1,40 @@
+/// Linearly interpolate `value` between `points`.
+///
+/// `points` must not be empty and contain unique x values sorted in ascending order.
+/// If `value` is smaller than the first point or larger than the last it is clamped.
+pub fn interpolate(value: f64, points: &[(f64, f64)]) -> f64 {
+    assert!(!points.is_empty());
+    assert!(points.windows(2).all(|window| window[0].0 < window[1].0));
+
+    if value < points[0].0 {
+        points[0].1
+    } else if let Some(window) = points
+        .windows(2)
+        .find(|window| value >= window[0].0 && value < window[1].0)
+    {
+        // https://en.wikipedia.org/wiki/Linear_interpolation#Linear_interpolation_between_two_known_points
+        let (x, x0, y0, x1, y1) = (value, window[0].0, window[0].1, window[1].0, window[1].1);
+        y0 + (x - x0) * ((y1 - y0) / (x1 - x0))
+    } else {
+        points.last().unwrap().1
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use assert_approx_eq::assert_approx_eq;
+
+    const LINEAR_DATA: &[(f64, f64)] = &[(1.0, 1.0), (2.0, 2.0), (3.0, 3.0)];
```

Comment:
I think a test case where the delta(0,1) != delta(1,2) would be interesting (and even what happens if the y values are not ascending).

Reply:
added